src/util/heap/gc_trigger.rs

```diff
@@ -139,27 +147,203 @@ pub struct MemBalancerTrigger {
     max_heap_pages: usize,
     /// The current heap size
     current_heap_pages: AtomicUsize,
+    /// Statistics
+    stats: AtomicRefCell<MemBalancerStats>,
+}
+
+#[derive(Copy, Clone, Debug)]
+struct MemBalancerStats {
+    // Allocation/collection stats in the previous estimation. We keep this so we can use them to smooth the current value
+    /// Previous allocated memory in pages.
+    allocation_pages_prev: f64,
+    /// Previous allocation duration in secs
+    allocation_time_prev: f64,
+    /// Previous collected memory in pages
+    collection_pages_prev: f64,
+    /// Previous colleciton duration in secs
+    collection_time_prev: f64,
+
+    // Allocation/collection stats in this estimation.
+    /// Allocated memory in pages
+    allocation_pages: f64,
+    /// Allocation duration in secs
+    allocation_time: f64,
+    /// Collected memory in pages
+    collection_pages: f64,
+    /// Collection duration in secs
+    collection_time: f64,
+
+    /// The time when this GC starts
+    gc_start_time: Instant,
+    /// The time when this GC ends
+    gc_end_time: Instant,
+
+    /// The live pages before we release memory.
+    gc_release_live_pages: usize,
+    /// The live pages at the GC end
+    gc_end_live_pages: usize,
+}
+
+/// A sentinel value for previous allocation/collection stats to indicate that no previous stats are recorded
+const NO_PREV: f64 = -1f64;
+
+impl std::default::Default for MemBalancerStats {
+    fn default() -> Self {
+        let now = Instant::now();
+        Self {
+            allocation_pages_prev: NO_PREV,
+            allocation_time_prev: NO_PREV,
+            collection_pages_prev: NO_PREV,
+            collection_time_prev: NO_PREV,
```

Comment:
The only use case of those fields (`smooth`) checks the value against `NO_PREV`. It is obviously a good use case of `Option<T>`.

```rust
        let smooth = |prev: Option<f64>, cur, factor| {
            prev.map(|p| p * factor + cur * (1.0f64 - factor)).unwrap_or(cur)
        };
```

Reply:
Done